dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java

```diff
@@ -0,0 +1,784 @@
+        final Register reg = ((ValueFunctionCall) call).getResultRegister();
+        return List.of(EventFactory.newLocal(reg, expressions.makeZero((IntegerType) reg.getType())));
+    }
+
+    private List<Event> inlinePthreadMutexLock(FunctionCall call) {
+        checkArgument(call.getArguments().size() == 1);
+        final Expression lockAddress = call.getArguments().get(0);
+        final String lockName = lockAddress.toString();
+        return List.of(EventFactory.Pthread.newLock(lockName, lockAddress));
+    }
+
+    private List<Event> inlinePthreadMutexUnlock(FunctionCall call) {
+        checkArgument(call.getArguments().size() == 1);
+        final Expression lockAddress = call.getArguments().get(0);
+        final String lockName = lockAddress.toString();
+        return List.of(EventFactory.Pthread.newUnlock(lockName, lockAddress));
+    }
+
+    private List<Event> inlineMalloc(FunctionCall call) {
+        if (call.getArguments().size() != 1) {
+            throw new UnsupportedOperationException(String.format("Unsupported signature for %s.", call));
+        }
+        final ValueFunctionCall valueCall = (ValueFunctionCall) call;
+        return List.of(EventFactory.newAlloc(
+                valueCall.getResultRegister(),
+                TypeFactory.getInstance().getByteType(),
+                valueCall.getArguments().get(0),
+                true
+        ));
+    }
+
+    private List<Event> inlineAssert(FunctionCall call) {
+        ExpressionFactory expressions = ExpressionFactory.getInstance();
+        final Expression condition = expressions.makeFalse();
+        final Event assertion = EventFactory.newAssert(condition, "user assertion");
+        final Event abort = EventFactory.newAbortIf(expressions.makeTrue());
+        abort.addTags(Tag.EARLYTERMINATION);
+        return List.of(assertion, abort);
+    }
+
+    // --------------------------------------------------------------------------------------------------------
+    // LLVM intrinsics
+
+    private List<Event> handleLLVMIntrinsic(FunctionCall call) {
+        assert call instanceof ValueFunctionCall && call.isDirectCall();
+        final ValueFunctionCall valueCall = (ValueFunctionCall) call;
+        final String name = call.getCalledFunction().getName();
+
+        if (name.startsWith("llvm.assume")) {
+            return List.of(EventFactory.newAssume(call.getArguments().get(0)));
+        } else if (name.startsWith("llvm.ctlz")) {
+            return inlineLLVMCtlz(valueCall);
+        } else if (name.startsWith("llvm.ctpop")) {
+            return inlineLLVMCtpop(valueCall);
+        } else if (name.contains("add.sat")) {
+            return inlineLLVMSaturatedAdd(valueCall);
+        } else if (name.contains("sub.sat")) {
+            return inlineLLVMSaturatedSub(valueCall);
+        } else if (name.startsWith("llvm.smax") || name.startsWith("llvm.smin")
+                || name.startsWith("llvm.umax") || name.startsWith("llvm.umin")) {
```

Comment:
why `startsWith` and not `equal`? Do they have type postfixes?

Reply:
Yes, see my reply above and why `*` was used.